### ISSUE TITLE
[css-flexbox-1] 9.2. Line Length Determination Issue

### ISSUE THREAD

**Issue by gitspeaks** [w3c/csswg-drafts#11392]
According to [CSS Flexible Box Layout Module Level 1, 9.2. Line Length Determination (3)(C)](https://www.w3.org/TR/css-flexbox-1/#algo-main-item):

> "If the used flex basis is content or depends on its available space, and the flex container is being sized under a min-content or max-content constraint (e.g. when performing automatic table layout [CSS21]), size the item under that constraint. The flex base size is the item’s resulting main size."

Problems:

1. It is unclear what constitutes a **"used flex basis value that depends on its available space"**, as none seem relevant in this context aside from the value `content`.

2. When a used `flex-basis` value is derived from a percentage that is resolved against an **indefinite containing block** (e.g., the flex container in this case), the resulting used value is `content`.

References:

1. According to [CSS Box Sizing Module Level 3](https://drafts.csswg.org/css-sizing-3/#indefinite):

   > "intrinsic sizing keywords such as max-content are indefinite"
   
2. According to [CSS Flexible Box Layout Module Level 1, Section 7.2.3](https://www.w3.org/TR/css-flexbox-1/#flex-basis-property):

   > "percentage values of flex-basis are resolved against the flex item’s containing block (i.e. its flex container); and if that containing block’s size is indefinite, the used value for flex-basis is content."

However, in practice, when a `flex-basis` used value of `content` is derived from a percentage resolved against an indefinite size flex container, **flex items are not consistently sized under the same constraint set on the flex container (e.g., `max-content`)**, as specified in 9.2(3)(c).

Test (Chrome and Firefox)

```html
<!DOCTYPE html>
<style>
  body,
  html {
    height: 100%;
    margin: 0;
  }
  .flex {
    display: flex;
    flex-flow: row nowrap;
    width: max-content;
    background: red;    
  }
  .flex > div {
    background: green;    
    /* flex-basis: content behaves as expected. */    
    flex-basis: 50%;
  }
</style>
<div class="flex">
  <div>1234 56789</div>
</div>
```

**Comment by gitspeaks** [comment]
Regarding point [4](https://www.w3.org/TR/css-flexbox-1/#algo-main-container) in the same section:

> Determine the main size of the flex container using the rules of the formatting context in which it participates. For this computation, auto margins on flex items are treated as 0.

It’s unclear why this step appears after calculating the flex base size and hypothetical main size of each item. Excluding the note about the automatic block size of a block-level flex container, this instruction seems more appropriate in point [2](https://www.w3.org/TR/css-flexbox-1/#algo-available), as the flex container's definite main size establishes the available main space for the flex items. Perhaps the intention here is to determine the main size of an indefinite flex container based on the sum of the hypothetical main sizes of its flex items?

**Comment by Loirooriol** [comment]
> It is unclear what constitutes a "used flex basis value that depends on its available space", as none seem relevant in this context aside from the value content.

There are things like `fit-content` and `stretch`.

> flex items are not consistently sized under the same constraint set on the flex container 

Your testcase doesn't show that. Flex items can re-resolve their percentages after the final size of the flex container is known.

**Comment by gitspeaks** [comment]
> There are things like fit-content and stretch.

How are these values applicable in this context?

The specification for [flex-basis](https://www.w3.org/TR/css-flexbox-1/#flex-basis-property) lists its possible values as `'content | <'width'>` with an initial value of `auto`. The  link  for [<'width'>](https://www.w3.org/TR/CSS21/visudet.html#propdef-width) which directs to the CSS 2.1 specification defines the valid values as:
`<length> | <percentage> | auto | inherit`.

>Your testcase doesn't show that. Flex items can re-resolve their percentages after the final size of the flex container is known.

What do you mean it doesn't show that?

Here is the render from the current version of Chrome:

![render](https://github.com/user-attachments/assets/e64ccc6c-a2fe-4686-b8cb-4203ea14dbfe)

Clearly, this is *not* resolved as `max-content`.

> Flex items can re-resolve their percentages after the final size of the flex container is known.

Are you suggesting that a flex container sized under `min-content` or `max-content` transitions from indefinite to definite size mid-algorithm, causing the layout algorithm to restart?

Where is this behavior specified in the [algorithm details](https://www.w3.org/TR/css-flexbox-1/#layout-algorithm)?

**Comment by Loirooriol** [comment]
You are looking at an old TR from 2018. Not that the draft is up-to-date with latest resolutions either, but at least if you checked https://drafts.csswg.org/css-flexbox-1/#flex-basis-property you would get a link to the modern syntax of `width`.

> Are you suggesting that a flex container sized under min-content or max-content transitions from indefinite to definite size mid-algorithm, causing the layout algorithm to restart?

Yes, that's kind of the idea. Please read about cyclic percentages in CSS Sizing.

**Comment by gitspeaks** [comment]
What do you mean I'm referencing an outdated TR from 2018? This is the [latest published version](https://www.w3.org/TR/css-flexbox-1/), and all the quotes in this issue are taken directly from it. Specifically, the issue regarding "available space" originates from this version.

Why isn't this considered a bug in the specification? While many things in *drafts* reflect improvements and enhancements, the current published version should, at the very least, be correct,  shouldn't it?

For example, why hasn't the link for width been updated? How can common reader make sense of things ? If the logic is to always goto the latest draft to verify things, than what's the value of the published version ?

**Comment by gitspeaks** [comment]
>Yes, that's kind of the idea. Please read about cyclic percentages in CSS Sizing.

So point [9.2(4)](https://www.w3.org/TR/css-flexbox-1/#algo-main-container) is effectively `goto` [9.2(2)](https://www.w3.org/TR/css-flexbox-1/#algo-available) when the flex container has an indefinite size  to make it definite ?

**Comment by Loirooriol** [comment]
Yeah the latest published thing is from 2018, so I'm calling it old.

> always goto the latest draft to verify things

I mean, not always. But if you want to discuss the spec details, then looking at an old thing isn't much productive.

> what's the value of the published version?

See https://www.w3.org/policies/process/#maturity-stages. In particular, editor's drafts are lacking in some areas:

> [Working Groups](https://www.w3.org/policies/process/#GroupsWG) and [Interest Groups](https://www.w3.org/policies/process/#GroupsIG) may make available [Editor’s drafts](https://www.w3.org/policies/process/#editors-draft). Editor’s drafts (ED) have no official standing whatsoever, and do not necessarily imply consensus of a [Working Group](https://www.w3.org/policies/process/#GroupsWG) or [Interest Group](https://www.w3.org/policies/process/#GroupsIG), nor are their contents endorsed in any way by W3C.

> So point [9.2(4)](https://www.w3.org/TR/css-flexbox-1/#algo-main-container) is effectively goto [9.2(2)](https://www.w3.org/TR/css-flexbox-1/#algo-available) when the flex container has an indefinite size to make it definite ?

Seems like that could be it. Not sure if it's clearly explained somewhere, I'm not that familiar with the flexbox spec, I know grid better (explained in https://drafts.csswg.org/css-grid/#layout-algorithm)

Note re-resolving percentages isn't specific of flexbox/grid, it's the same for the contents of an inline-block (in the inline axis only).

**Comment by gitspeaks** [comment]
>I mean, not always. But if you want to discuss the spec details, then looking at an old thing isn't much productive.

It's not just about discussing the details — it's about pointing out errors in the current version to ensure they get 'patched' For example, shouldn't the incorrect `width` link be fixed? What's the policy regarding the correctness of the existing published version?

> Seems like that could be it. Not sure if it's clearly explained somewhere, I'm not that familiar with the flexbox spec, I know grid better (explained in https://drafts.csswg.org/css-grid/#layout-algorithm)

So, who actually knows? And who should be asked about this?

**Comment by gitspeaks** [comment]
@tabatkins  Is point [9.2(4)](https://www.w3.org/TR/css-flexbox-1/#algo-main-container) effectively a reference back to [9.2(2)](https://www.w3.org/TR/css-flexbox-1/#algo-available)  for the case when the flex container has an indefinite size, to make it definite after the flex items main sizes have been determined?

**Comment by Loirooriol** [comment]
> shouldn't the incorrect width link be fixed?

It's not incorrect. It's just that the definition of `width` in CSS Sizing supersedes the one from CSS2.

> So, who actually knows? And who should be asked about this?

You can file an issue asking for this to be clarified, like #3418 did for grid.

**Comment by gitspeaks** [comment]
> It's not incorrect. It's just that the definition of width in CSS Sizing supersedes the one from CSS2.

I disagree. In the context of this issue (e.g *published version*), the provided **link** for `width` in [7.2.3](https://www.w3.org/TR/css-flexbox-1/#flex-basis-property) is incorrect.

I initially asked:


> 1. It is unclear what constitutes a **"used flex basis value that depends on its available space"**, as none seem relevant in this context aside from the value `content`.

You replied:


> There are things like `fit-content` and `stretch`.

To which I responded:

> The specification for [flex-basis](https://www.w3.org/TR/css-flexbox-1/#flex-basis-property) lists its possible values as `'content | <'width'>` with an initial value of `auto`. The link for [<'width'>](https://www.w3.org/TR/CSS21/visudet.html#propdef-width) which directs to the CSS 2.1 specification defines the valid values as:
> `<length> | <percentage> | auto | inherit`.


Given this, how can one interpret "used flex basis value that depends on its available space" as anything other than `content` without prior knowledge that `fit-content` and `stretch` are relevant values?

Possible solutions:

1) Update the link to point to a more current and relevant definition of `width`.

2) Explicitly include `fit-content` and `stretch` in the list of valid flex-basis values.

**Comment by gitspeaks** [comment]
What's the reason for the downvote?

I repeat:

> According to [CSS Flexible Box Layout Module Level 1, 9.2. Line Length Determination (3)(C)](https://www.w3.org/TR/css-flexbox-1/#algo-main-item):
> 
> > "If the used flex basis is content or depends on its available space, and the flex container is being sized under a min-content or max-content constraint (e.g. when performing automatic table layout [CSS21]), size the item under that constraint. The flex base size is the item’s resulting main size."
> 

It's unclear what "or depends on its available space" means here, given that the possible values for flex-basis are limited to `content` and the `width` as defined in CSS2.

You pointed out that the definition of `width` was [extended](https://drafts.csswg.org/css-sizing-3/#propdef-width) in draft and the new values in [css-sizing-4](https://drafts.csswg.org/css-sizing-4/#sizing-values) . I'm simply suggesting backporting a link to these definitions into the published version of the spec for clarity. Why the pushback?
